libs/MobileSync/MobileSyncTests/SyncUpTargetTests.m

```diff
@@ -108,6 +108,22 @@ -(void) testSyncUpWithErrors
     [idToFields addEntriesFromDictionary:idToFieldsGoodNames];
 }
 
+/**
+ * Sync up records missing sobject type
+ */
+- (void) testSyncUpWithNoType {
```

Comment:
New tests:
- no type
- bad typ
Those tests run for both SyncUpTarget and BatchSyncUpTarget.
Before the code change, the test would repro the infinite looping.